cpp/src/parquet/encoding.cc

```diff
@@ -3037,11 +3073,179 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+      PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
+                                                                   this->memory_pool()));
+      T* data = reinterpret_cast<T*>(buffer->mutable_data());
+      int num_valid_values = ::arrow::util::internal::SpacedCompress<T>(
+          src, num_values, valid_bits, valid_bits_offset, data);
+      Put(data, num_valid_values);
+    } else {
+      Put(src, num_values);
+    }
+  }
+
+ protected:
+  template <typename ArrayType>
+  void PutBinaryArray(const ArrayType& array) {
+    uint32_t previous_len = static_cast<uint32_t>(last_value_.size());
+    std::string_view last_value_view = last_value_;
+
+    PARQUET_THROW_NOT_OK(::arrow::VisitArraySpanInline<typename ArrayType::TypeClass>(
+        *array.data(),
+        [&](::std::string_view view) {
+          if (ARROW_PREDICT_FALSE(view.size() >= kMaxByteArraySize)) {
+            return Status::Invalid("Parquet cannot store strings with size 2GB or more");
+          }
+          // Convert view to ByteArray so it can be passed to the suffix_encoder_.
+          const ByteArray src{view};
+
+          uint32_t j = 0;
+          while (j < std::min(previous_len, src.len)) {
+            if (last_value_view[j] != view[j]) {
+              break;
+            }
+            j++;
+          }
+          previous_len = src.len;
+          prefix_length_encoder_.Put({static_cast<int32_t>(j)}, 1);
+
+          const uint8_t* suffix_ptr = src.ptr + j;
+          const uint32_t suffix_length = static_cast<uint32_t>(src.len - j);
+          last_value_view = view;
+          // Convert suffix to ByteArray so it can be passed to the suffix_encoder_.
+          const ByteArray suffix(suffix_length, suffix_ptr);
+          suffix_encoder_.Put(&suffix, 1);
+
+          return Status::OK();
+        },
+        []() { return Status::OK(); }));
+    last_value_ = last_value_view;
+  }
+
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> prefix_length_encoder_;
+  DeltaLengthByteArrayEncoder<ByteArrayType> suffix_encoder_;
+  std::string last_value_;
+};
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const T* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+  ArrowPoolVector<int32_t> prefix_lengths(num_values,
+                                          ::arrow::stl::allocator<int32_t>(pool_));
+  std::string_view last_value_view = last_value_;
+
+  int i = 0;
+  while (i < num_values) {
+    // Convert to ByteArray so we can pass to the suffix_encoder_.
+    auto value = reinterpret_cast<const ByteArray*>(&src[i]);
+    if (ARROW_PREDICT_FALSE(value->len >= kMaxByteArraySize)) {
+      throw Status::Invalid("Parquet cannot store strings with size 2GB or more");
+    }
+
+    auto view = string_view{reinterpret_cast<const char*>(value->ptr), value->len};
+    uint32_t j = 0;
+    while (j < std::min(value->len, static_cast<uint32_t>(last_value_view.length()))) {
```

Comment:
Maybe the compiler is smarter enough to optimize this, but I wonder if we should pull out this cast and comparison from the loop condition?

There's a similar case above.

```suggestion
    int32_t common_length = std::min(value->len, static_cast<uint32_t>(last_value_view.length()));
    while (j < common_length) {
```

Reply:
It's definetly better for readability!

Reply:
Ah I see now, `last_value_view` was being mutated in the loop.

Reply:
Yeah, that seems to be the case.

Reply:
Ah, but it's still not fixed. I'll take a look this tomorrow.